accepted/0000-workspaces.md

```diff
@@ -0,0 +1,479 @@
+- `test` Run tests in all packages
+- `update` Update a dependency in all packages
+- `version` Bump versions for all packages
+- `view` View registry info for all packages
+
+A new config value should be introduced in order to allow for filtering out a subset of the packages in which to run these commands. e.g: `--filter`
+
+#### Test example:
+
+```
+├── package.json { "name": "foo", "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "version": "1.0.0" }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm test
+
+> foo@1.0.0 test /Users/username/foo
+> echo "Error: no test specified" && exit 1
+
+Error: no test specified
+npm ERR! Test failed.  See above for more details.
+
+> dep-a@1.0.0 test /Users/username/foo/dep-a
+> done
+
+> dep-b@1.3.1 test /Users/username/foo/dep-b
+> done
+```
+
+
+### 4. Publishing workspaces
+
+A workspace may not be published to the registry and for all publishing purposes having a valid `"workspace"` entry in a `package.json` is going to be the equivalent of `"private": true`.
+
+
+## Examples
+
+### 1. Simplistic example expanding on symlinking structure and `package-lock` file shape.
+
+Given a workspaces setup with the following contents:
+
+```
+$ cat ./package.json
+{
+    "name": "foo",
+    "version": "1.0.0",
+    "workspaces": [
+        "./core/*",
+        "./packages/*"
+    ],
+    dependencies: {
+        "lodash": "^4.x.x",
+        "libnpmutil": "^1.0.0"
+    }
+}
+
+$ cat ./core/libnpmutil/package.json
+{
+    "name": "libnpmutil",
+    "version": "1.0.0",
+    "dependencies": {
+        "lodash": "^4.x.x"
+    }
+}
+
+$ cat ./packages/workspace-a/package.json
+{
+    "name": "workspace-a",
+    "version": "1.7.3",
+    "peerDependencies": {
+        "react": "^16.x.x"
+    },
+    "dependencies": {
+        "workspace-b": "^2.0.0"
+    }
+}
+
+$ cat ./packages/worskpace-b/package.json
+{
+    "name": "workspace-b",
+    "version": "2.1.1",
+    "peerDependencies": {
+        "react": "^16.x.x"
+    }
+}
+```
+
+Will result in the following symlinking structure:
+
+```
+$ tree
+.
+├── package-lock.json
+├── node_modules
+│   ├── lodash
+│   ├── libnpmutil -> ./core/libnpmutil
+│   ├── workspace-a -> ./packages/workspace-a
+│   ├── workspace-b -> ./packages/workspace-b
+│   └── react
+├── core
+│   └── libnpmutil
+│       ├── package-lock.json
+│       └── node_modules
+│           └── lodash -> ../../../node_modules/lodash
+└── packages
+    ├── workspace-a
+    │   ├── package-lock.json
+    │   └── node_modules
+    │       ├── react -> ../../../node_modules/react
+    │       └── worspace-b -> ../../../node_modules/workspace-b
+    └─ worspace-b
+        ├── package-lock.json
+        └── node_modules
+            └── react -> ../../../node_modules/react
+```
+
+And the following `package-lock.json` files:
+
+```
+$ cat ./package-lock.json
+{
+  "name": "foo",
+  "version": "1.0.0",
+  "lockfileVersion": 2,
+  "requires": true,
+  "packages": {
+    "": {
```

Comment:
Can we name this something more significant, like ... `_rootManifest`? An empty string as a key feels like high potential for long-term regret.

Reply:
oh this is outside of the scope of this RFC, this is just a basic `package-lock` v2 template - maybe it's worth poking the team at https://github.com/npm/arborist if you have strong feelings about it 😊

Reply:
Lockfiles are for machines.  The key in the `packages` section of a v2 or v3 lockfile is the `/`-normalized relative path from the root to the node in question.  There's a valid point as to whether it ought to be in the lockfile at all, since it's duplicative with `package.json`, but if it _is_ included, then `""` is the most portable and sensible key for it, imo.  Having to constantly do `node.location || '_rootManifest'` (and then transpose it back, as well) imposes a fair bit of added complexity in the metadata handling logic.